### ISSUE TITLE
Not possible to set SO_KEEPALIVE for Socket

### ISSUE THREAD

**Issue by pheki** [dart-lang/sdk#45125]
In the current dart SDK (2.10.5), enabling or disabling TCP Socket KEEPALIVE is not available via Socket.setSocketOption (which uses `RawSocketOption`).

Ideally, keepalive period (TCP_KEEPIDLE / TCP_KEEPALIVE) /  could also be exposed, as its supported on both iOS + Android (but with different options)

Is this planned?

Thanks for your attention.

**Comment by mraleph** [comment]
It is supported. You just need to construct `RawSocketOption` with the right `level` and `option`. For example on iOS and Mac OS that would be `RawSocketOption.fromBool(/* SOL_SOCKET */ 0xffff, /* SO_KEEPALIVE */ 0x0008, true)` and `RawSocketOption.fromInt(/* IPPROTO_TCP */ 6, /* TCP_KEEPALIVE */ 0x10, secs)` to enable keepalive and set keepalive period. 

Have you tried this? Does it not work for you for some reason?

**Comment by mraleph** [comment]
For Android you would have to use `RawSocketOption.fromBool(/* SOL_SOCKET */ 0x1, /* SO_KEEPALIVE */ 0x0009, true)` and `RawSocketOption.fromInt(/* IPPROTO_TCP */ 6, /* TCP_KEEPIDLE */ 4, secs)` if I am reading headers correctly.

**Comment by pheki** [comment]
I didn't actually try it as I wasn't finding the correct headers and though that using hard-coded values wasn't a great idea (but thinking about it now, its not a problem since the platform can't change the value as it mean losing compatibility).

I tried both (and some other options) in both iOS and Android and it worked fine.

Thanks a lot!

**Comment by MahmoodBakhshayesh** [comment]
can someone help me do that too? 
i'm trying to set a socket keep-alive so when socket or connection disconnected i got notified
i added the rawSocketOption but when i disconnect phone from internet nothing happens
can you show me in code and how would i know that connection is lost?
does it call an Exception or onDone?

**Comment by pheki** [comment]
Hey @MahmoodBakhshayesh, maybe the default interval is too high?

I believe it will throw on `socket.done`, so `socket.done.catchError((error) { })` would catch it.

This is the code I'm using to enable keepalive on both iOS and Android with configurable interval:

```dart
void _enableKeepalive(Socket socket) {
  // Enable keepalive probes every 60 seconds with 3 retries each 10 seconds
  final keepaliveEnabled = true;
  final keepaliveInterval = 60;
  final keepaliveSuccessiveInterval = 10;
  final keepaliveSuccessiveCount = 3;

  if (Platform.isIOS || Platform.isMacOS) {
    final enableKeepaliveOption = RawSocketOption.fromBool(
      0xffff, // SOL_SOCKET
      0x0008, // SO_KEEPALIVE
      keepaliveEnabled
    );
    final keepaliveIntervalOption = RawSocketOption.fromInt(
      6, // IPPROTO_TCP
      0x10, // TCP_KEEPALIVE
      keepaliveInterval
    );
    final keepaliveSuccessiveIntervalOption = RawSocketOption.fromInt(
      6, // IPPROTO_TCP
      0x101, // TCP_KEEPINTVL
      keepaliveSuccessiveInterval,
    );
    final keepaliveusccessiveCountOption = RawSocketOption.fromInt(
      6, // IPPROTO_TCP
      0x102, // TCP_KEEPCNT
      keepaliveSuccessiveCount
    );

    socket.setRawOption(enableKeepaliveOption);
    socket.setRawOption(keepaliveIntervalOption);
    socket.setRawOption(keepaliveSuccessiveIntervalOption);
    socket.setRawOption(keepaliveusccessiveCountOption);
  } else if (Platform.isAndroid) {
    final enableKeepaliveOption = RawSocketOption.fromBool(
      0x1, // SOL_SOCKET
      0x0009, // SO_KEEPALIVE
      keepaliveEnabled
    );
    final keepaliveIntervalOption = RawSocketOption.fromInt(
      6, // IPPROTO_TCP
      4, // TCP_KEEPIDLE
      keepaliveInterval
    );
    final keepaliveSuccessiveIntervalOption = RawSocketOption.fromInt(
      6, // IPPROTO_TCP
      5, // TCP_KEEPINTVL
      keepaliveSuccessiveInterval,
    );
    final keepaliveusccessiveCountOption = RawSocketOption.fromInt(
      6, // IPPROTO_TCP
      6, // TCP_KEEPCNT
      keepaliveSuccessiveCount
    );

    socket.setRawOption(enableKeepaliveOption);
    socket.setRawOption(keepaliveIntervalOption);
    socket.setRawOption(keepaliveSuccessiveIntervalOption);
    socket.setRawOption(keepaliveusccessiveCountOption);
  }
}
```

**Comment by Hruthik0x** [comment]
@pheki  Thank you very much I needed these options real bad. 
Nothing is mentioned about these options in the docs, GPT, Copilot everywhere its mentioned dart does not support these options. 
Me and my friend (@DeepakJhawar) were struggling for 3 days straight to integrate Java and Kotlin (as they support these features) and  we failed.
Thank you very much, now we simply used the code you mentioned, no pain of writing in Java or Kotlin and integrating it.
